### ISSUE TITLE
How to handle legacy Ruby content that may use <rbc>

### ISSUE THREAD

**Issue by travisleithead** [w3c/html#291]
Copied from Bugzilla: https://www.w3.org/Bugs/Public/show_bug.cgi?id=26426

@kojiishi, @darobin, @fantasi, @r12a 

> Tree construction, The ["in body" insertion mode](http://w3c.github.io/html/syntax.html#the-in-body-insertion-mode) states that:
> 
> A start tag whose tag name is one of: "rb", "rp", "rtc"
> If the stack of open elements has a ruby element in scope, then generate implied end tags. If the current node is not then a ruby element, this is a parse error.
> 
> This rule could break existing HTML + XHTML Ruby annotation content because they could use rbc tags:
> 
> ``` html
> <ruby><rbc><rb>base</rb></rbc>...
> ```
> 
> Such content will be a parse error because rb is not a direct child of ruby.
> ## Should we loosen the rule to require rb being the direct child of ruby?
> 
> [tag] [reply] [−]  Comment 1  fantasai     2014-07-24 11:02:41 UTC  
> 
> I think
> 
> ```
>  * <rb> should only generate implied end tags for <rb>,<rt>,<rtc>,<rp>
>  * <rt> should only generate implied end tags for <rb>,<rt>,<rtc>,<rp>
>  * <rtc> should only generate implied end tags for <rb>,<rt>,<rtc>,<rp>
>  * <rp> should not generate implied end tags for anything (bug 26424).
> ```
> ---
> 
> [tag] [reply] [−]  Comment 2  Koji Ishii     2014-07-26 12:40:45 UTC  
> 
> Changing auto-closing rules doesn't help this. Because rbc is not defined in HTML5, we can't define its auto-closing rules.
> 
> In other words, I could say following fragments exist in the wild:
> 
> ``` html
> <ruby><any><rb>base</rb></any>...</ruby>
> ```
> 
> and the `<any>` is an undefined element. If this fragment is defined as a parser error, we're likely to break a lot of existing documents that follow XHTML Ruby Annotation spec.
> ---
> 
> [tag] [reply] [−]  Comment 3  Koji Ishii     2014-07-28 18:42:33 UTC  
> 
> After some more tests, I confirmed that having `<rbc>` doesn't render as proper ruby even before the spec change, so I'd put lower priority.
> 
> Gecko, after the spec change was implemented, starts showing this as a parser error in Source view because `<rb>` is now defined, as discussed in https://bugzilla.mozilla.org/show_bug.cgi?id=1042885

**Comment by chaals** [comment]
Check if we match HTML 5, and what we need to do…

**Comment by chaals** [comment]
From [discussion at last week's face to face meeting](https://www.w3.org/2017/03/28-html-minutes.html#item10):

Proposal is to remove ruby entirely from the main spec, and update the [Ruby Annotations Recommendation](https://www.w3.org/TR/ruby/) with a new version that obsoletes the various things that are out there. Note that there is also an [editor's draft for a Ruby extension](http://darobin.github.io/html-ruby/).

Keeping this open until that process is set in motion.

**Comment by r12a** [comment]
This issue is about `rbc`, which as @kojiishii points out, is not part of the HTML content model.  It would certainly be worth considering whether that should or should not at some point be brought into the HTML content model, but that's another issue altogether. As, i think, is the question of whether or not to remove ruby entirely from the HTML spec, and a separate issue should be raise for that discussion (not least, so that people can find it again some time in the future).

In the meantime, let me add a few thoughts, hopefully reasonably coherent as i try to deal with the shock of hearing a proposal to remove ruby from HTML...

Multiple discussions over past years with Japanese web and epublishing communities point to support for ruby and vertical text as their most pressing needs. As a result we spent a significant effort over the past few years on the ruby content model for the HTML spec and getting browser support for it.

It may indeed be useful to update the content model for ruby in the Ruby Annotation spec (although i don't know who's going to do that, or how long will it take), but  I can't see how removing ruby from HTML in the meantime helps anyone.

Note, btw, that an update of the Ruby Annotation spec should imho provide a generic content model (and not syntactic detail) that can be applied to other scenarios than just HTML. The i18n WG has had to deal with requests for advice on such from groups working on  ITS, WebVTT, and TTML, and i'm sure it would be of use to other people working with XML or non-XML formats.

Some clarifications for rtc, since the minutes of the FTF say that "We only have ruby element in HTML, never got rtc supported", let me say this: As i understand it, **some considerable time ago the major browsers already provided support for parsing the ruby content model in HTML, including rtc**. What not all of them do at the moment (though Firefox does) is provide useful default display positioning for markup involving rtc, but rendering is not required by the HTML spec - it's focus is the content model for ruby related items and syntactic model that handles the markup. My understanding is that the browser implementers are just waiting on the CSS Ruby spec for the rendering, and that in fact the CSS Ruby spec will give implementers good reasons to provide better support for rtc rendering.

By the way, i believe that the content of the editor's draft for a ruby extension pointed to in the previous comment was merged into the HTML5 spec in order to provide the current HTML text, and it's no longer intended to be useful. It would perhaps be helpful to point that out in the extension spec itself.

@fantasai @kojiishi @darobin thoughts?

**Comment by chaals** [comment]
@r12a I also understand that all the pieces of `ruby` in HTML except `rtc` rendering are interoperably implemented. 

The idea is not to argue that ruby somehow isn't part of HTML. Indeed, the basis for doing the work on updating the old [Ruby Annotation Recommendation](http://w3.org/tr/ruby) is that it **is** part of HTML, which is why it is in scope.

Instead, the point of the proposal is to break it out as one of the [modules that extend HTML](https://github.com/w3c/html-extension) and make a first-class Recommendation that explains how ruby integrates into HTML. My instinct is that this is compatible with it explaining how to do Ruby in other places, but if it isn't, that would be a reason to change the proposal. Likewise, a risk of the Ruby update somehow leaving ruby itself in the cold for a long time would be a reason to change the proposal.

It is just a proposal.

**Comment by aphillips** [comment]
This was discussed in the I18N teleconference this past [week](https://lists.w3.org/Archives/Public/www-international/2017AprJun/0006.html) and I drew an [action item](https://www.w3.org/International/track/actions/609) to write back to @chaals and HTML regarding this.

Our WG would like to request that:

1. Ruby not be removed from the core HTML spec until the replacement documents are ready.
2. When text is removed, we would like to review the results.
3. We intend to create a "2.0" version of [Ruby](https://www.w3.org/TR/ruby/) which will describe the general model and requirements for Ruby (such as @r12a describes above). This document will probably be much simpler than the current Ruby rec.
4. Either HTML or I18N can create an HTML Ruby extension specification. Ideally this would depend on Ruby2. This specification would make clear that it is integrated to HTML. Work on this could begin earlier than Ruby2 being finalized, as long as the integral relationship to HTML is clear.

How do we proceed? Does HTML want to own the extension spec? Or should I18N own it?

**Comment by siusin** [comment]
Thanks @aphillips !

All of these points make perfect sense to me. It'd be great if the I18N could co-edit the HTML Ruby extension spec, perhaps a joined deliverable?

**Comment by siusin** [comment]
The spec has been updated to include expection for `<rtc>`:

> A start tag whose tag name is one of: "rb", "rtc"
> If the stack of open elements has a ruby element in scope, then generate implied end tags. If the current node is not now a ruby element, this is a parse error.
> 
> Insert an HTML element for the token.
> 
> A start tag whose tag name is one of: "rp", "rt"
> If the stack of open elements has a ruby element in scope, then generate implied end tags, except for rtc elements. If the current node is not now a rtc element or a ruby element, this is a parse error.
> 
> Insert an HTML element for the token.

1] `<rb>`, `<rt>`, `<rp>` generate implied end tags in `<ruby>`
2] `<rt>`generate implied end tags in `<rtc>`
3] `<rb>` generate implied end tags in `<any></any>`
4] `<rb>` generate implied end tags in `<any>`
FF, SF, Ch implementations follow 1];
FF implementation follows 2];
FF, SF, Ch implementations follow 3];
FF implementation follows 4].

**Comment by LJWatson** [comment]
From the [HTML meeting today](https://www.w3.org/2018/04/24-html-minutes.html) it seems that ```<rtc>``` and ```<rt>``` may not have sufficient implementation. We should mark these as "at risk" until we can run some tests to confirm/deny.

**Comment by siusin** [comment]
@LJWatson sorry I wasn't clear in the minute, `<rtc>` only gets [one implementation](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/rtc), and should be mard 'at-risk'. `<rt>` is fine.

Also see test for [Ruby implied end tags](https://rawgit.com/siusin/web-platform-tests/ruby-implied-end/html/semantics/text-level-semantics/the-ruby-element/ruby-implied-end.html).

**Comment by siusin** [comment]
More Ruby tests from the i18n WG:

- https://w3c.github.io/i18n-tests/results/ruby-html
- https://www.w3.org/International/tests/#html5-devt
